### ISSUE TITLE
AnyPointer field defaults are admitted by the compiler, but ignored by the c++ implementation.

### ISSUE THREAD

**Issue by popham** [capnproto/capnproto#633]
I assume that default `AnyPointer` field values will be supported under c++ eventually. Have you put any thought into the eventual behavior? I think I see how it has gotta be, but I don't especially like it. What's bugging me is copying or moving a null AnyPointer field with no default value to an AnyPointer field with a default value, e.g.

```
struct K {
  id @0 :Int32;
}
const someK :K = (id=13);
struct Org {
  anAny @0 :AnyPointer;
  anotherAny @1 :AnyPointer = .someK;
}
```

I've convinced myself that copying or moving a null `anAny` field to an `anotherAny` field should just null the `anotherAny` pointer and expose `someK` on the next dereference of `anotherAny`'s pointer. I've also convinced myself that the analogous situation with `AnyStruct` should write an empty struct pointer (offset=-1); this feels like a warty API. More bad news, I'm 90% sure that there is no empty list pointer analogue to the empty struct pointer, meaning that `AnyList` must behave like `AnyPointer` instead of `AnyStruct`.

Reading what I've written, it looks to me like I should drop the offset=-1 `AnyStruct`. Or should `AnyPointer`, `AnyStruct`, and `AnyList` all refuse default values? Or should offset=-1 analogues be introduced for lists and `AnyPointer`s?

P.S. Here's code establishing my claim that the c++ implementation is ignoring default values for AnyPointer fields if you're skeptical:

```
#include "./def.capnp.h"
#include <capnp/message.h>
#include <capnp/serialize-packed.h>
#include <iostream>

int main() {
  ::capnp::MallocMessageBuilder message;

  Org::Builder org = message.initRoot<Org>();
  auto an = org.getAnAny();
  auto another = org.getAnotherAny();

  std::cout << "id of `anAny`: " << an.getAs<K>().getId() << std::endl;
  std::cout << "id of `anotherAny`: " << another.getAs<K>().getId() << std::endl;

  if (!an.isNull()) {
    std::cout << "The `anAny` field is non-null" << std::endl;
  }
  if (!another.isNull()) {
    std::cout << "The `anotherAny` field is non-null" << std::endl;
  }

  return 0;
}
```

**Comment by kentonv** [comment]
To review, here's how default values for pointer fields behave:

1. If you call the getter for a pointer field on a Reader, the returned value is a Reader for a shared read-only copy of the default value.
2. If you call the getter for a pointer field on a Builder, then the field is just-in-time initialized to a copy of its default value, then the Builder for the field is returned as normal. If you don't want to initialize to a copy of the default value, you must use `initFoo()` instead of `getFoo()`.

I'm not super-proud of these rules, but that's how the C++ implementation does it.

I can certainly believe that AnyPointers with defaults are broken in the C++ implementation. The ability to specify a literal value for an AnyPointer was added much later and was primarily intended for use in constants. I may have forgotten about default values when I added it.

> I've convinced myself that copying or moving a null anAny field to an anotherAny field should just null the anotherAny pointer and expose someK on the next dereference of anotherAny's pointer.

I think you're right. AnyPointer is unique among types in that it has no "default default" value. It's "default default" is simply "null".

Perhaps if I ever get around to implementing `Maybe(T)`, then we can say that an `AnyPointer` field is required either to be `Maybe(AnyPointer)` or to have an explicit default value. (I assume the vast majority of use cases would use `Maybe(AnyPointer)`.

> I've also convinced myself that the analogous situation with AnyStruct should write an empty struct pointer (offset=-1); this feels like a warty API.

Yes, because the "default default" for a struct is the empty struct. So the getter returns a Reader for the empty struct and the setter assigns that into `anotherAny`. This seems like the intuitively-correct behavior to me; `anotherAny` ends up containing the same value as `anAny`.

This is a little quirky, though: The "default default" for `AnyStruct` has no specific struct type. This just happens to work because struct types are entirely erased from the wire format.

> More bad news, I'm 90% sure that there is no empty list pointer analogue to the empty struct pointer, meaning that AnyList must behave like AnyPointer instead of AnyStruct.

The "default default" for a list is a list with zero size. However, I suppose in the case of `AnyList` this is underspecified: we need to know not just the size, but also the element size (type). I suppose one possibility would be to assume `Void`, though that feels arbitrary.

Perhaps the "must have default or be Maybe" rule should apply to AnyStruct and AnyList as well.

Note that no such rule is necessary for `Capability` (which is basically `AnyInterface`), since the only default for any capability type is the null capability whose methods all throw exceptions...

**Comment by popham** [comment]
Have you considered a new pointer encoding for empty list, empty AnyPointer, and possibly even empty struct? I anticipate `Maybe` falling into a pair of cases:

* Only allow `Maybe` for struct field types. This will allow a corresponding flag to be encoded in the data section, and `Maybe` qualification will only cost 1 bit.
* Allow `Maybe` outside of struct field typing. This will cost an additional word.

I suspect that the latter case is the target. If so, then a new pointer encoding may be a good option. The argument for abandoning empty structs as currently implemented points out how 0x00000000fffffffc doesn't pack as small as, say, 0x000000000000000f.

**Comment by popham** [comment]
So if I have a `Maybe(AnyPointer)` typed field, then I can't use its value to naively set a default-valued `AnyPointer` field? Only "Just" cases can set a default-valued `AnyPointer` field, whereas I have to improvise in the "Nothing" case (even if that improvisation is just nulling the default-valued `AnyPointer` field)? I love it (I think).

In the opposite direction the `Maybe(AnyPointer)` typed field always receives non-null values from a default-valued `AnyPointer` field?

From a plugin authoring perspective, I'm going to reject schemas with default-valued `AnyX` fields for now.

**Comment by kentonv** [comment]
> In the opposite direction the Maybe(AnyPointer) typed field always receives non-null values from a default-valued AnyPointer field?

Right, I think an `AnyPointer::Reader` and `AnyPointer::Builder` would no longer be permitted to represent a null AnyPointer value, hence setting a `Maybe<AnyPointer>` from either of these would necessarily set it to a non-null value.
